packages/agents/sequencer/src/lib/operations/auctions.ts

```diff
@@ -170,10 +190,55 @@ export const executeAuctions = async (_requestContext: RequestContext) => {
         let taskId: string | undefined;
         // Try every bid until we find one that works.
         for (const randomBid of randomized) {
+          // Sanity: Check if this router has enough funds.
+          const { router } = randomBid;
+          const { amount: _amount, local: asset } = bidData;
+          const amount = BigNumber.from(_amount);
+          let routerLiquidity: BigNumber | undefined = await cache.routers.getLiquidity(router, destination, asset);
+
+          if (!routerLiquidity) {
+            // Either we haven't cached the liquidity yet, or the value cached has become expired.
+            routerLiquidity = await subgraph.getAssetBalance(destination, router, asset);
+            if (routerLiquidity) {
+              await cache.routers.setLiquidity(router, destination, asset, routerLiquidity);
```

Comment:
Is it dangerous to have this value cached vs querying the subgraph every time? Subgraphs do a lot of caching themselves. I'm a little worried this value is not going to get properly decremented when the liquidity is removed after an execution and the cache will become out of sync.

Reply:
As soon as we send off a single execute, it will become out of sync

Reply:
We are rapid-fire looping through these txs. When an `execute` is sent off (to the relayer), it is never sent+accepted right away. Even when it is accepted, the amount of liquidity difference won't be confirmed for another few blocks.

I'm trying to avoid double-spend of the router's funds. Although technically speaking, it likely wouldn't end in reverts here. It would end in the relayer doing nothing, and then we need track what the relayer is doing, and then we need to re-queue the transfer, etc etc etc

This way things are "pessimistic" in that we assume the tx will go thru and the router will be debited.

Reply:
I think of the subgraph's record as a lagging indicator of router liquidity

Reply:
Regardless, I think you might be right in another sense... Ima have router liquidity expire a lot faster, and remove the approval stuff for now (I kri... my unit tests, I worked so hard....)